pd2pg.rb

```diff
@@ -123,32 +175,46 @@ def convert_incident(i)
   # values from the API. Yields each API value to a block that should
   # convert the API value to a DB record for subsequent insertion /
   # update.
-  def refresh_bulk(collection)
+  def refresh_bulk(collection, table_name=nil, endpoint=nil, additional_headers={}, should_log=true)
     # Fetch all values from the API and apply the conversion block to
     # each, forming an in-memory array of DB-ready records.
+    if endpoint.nil?
+      endpoint = collection
+    end
+
+    if table_name.nil?
+      table_name = collection
+    end
+
     offset = 0
     total = nil
     records = []
     while !total || offset <= total
-      log("refresh_bulk.page", collection: collection, offset: offset, total: total || "?")
+      if should_log
+        log("refresh_bulk.page", table_name: table_name, offset: offset, total: total || "?")
+      end
       response = api.request(
         :method => :get,
-        :path => "/api/v1/#{collection}",
-        :query => {"offset" => offset, "limit" => PAGINATION_LIMIT},
+        :path => "/api/v1/#{endpoint}",
+        :query => {"offset" => offset, "limit" => PAGINATION_LIMIT}.merge(additional_headers),
         :expects => [200]
       )
       data = JSON.parse(response.body)
-      total = data["total"]
+      total = data["total"] || data[collection.to_s].length
       offset = offset + PAGINATION_LIMIT
       items = data[collection.to_s]
       records.concat(items.map { |i| yield(i) })
     end
+    if should_log
+      log("refresh_bulk.update", table_name: table_name, total: records.length)
+    end
+    database_update(table_name, records)
```

Comment:
with the addition of this, is the `Yields each API value to a block` above still true?

Reply:
Yes. It's kind of hard to see, but the `database_update` method is code I copypasted from the bottom of this method into its own method. So it's pretty much the same behavior, but "decomposed". The yield is still used!